### ISSUE TITLE
checkpoint merge & model averaging via dummy optimizer

### ISSUE THREAD

**Issue by rakkit** [pytorch/torchtitan#3982]
I mentioned about EMA checkpoint (or model averaging) times ago here https://github.com/pytorch/torchtitan/issues/2355  and i dont see any related updates or new pr try to support it in torchtitan.

and i saw there is multiple optimizers support PR is merged https://github.com/pytorch/torchtitan/pull/3269 and i wonder that why dont we just add a dummy optimizer and make it state's as EMA weights? so all parallel stuff can essentially resolved automatically, and the core code is also gonna be simple, we can even borrow adamw's fused update there, and even cpu offload stuff.  and async ckpt saving can also handle the checkpoint saving ...   even though this solution maybe tricky to do smth like run val loss on EMA weightsetc but the solution itself is very easy to go.

i wonder like to discusses in brief see if this is a good idea to do then i can open PR (i think this should be easy to do). 

----

about why we are interested in model averaging.  so most of now day's llm trianing go with stable-decay LR scheduler, and one problem is, you cannot see model's actually performance in stable LR regime on-real-time.   (so you see model performance hugely improved in decay phase) and this is annoying.    and model averaging is the solution for that.   In stable-LR phase, instead of checking eval on current weights,  the averaged EMA weights can better measure the model's performance. 

And the extension of it will be scheduler-free, so one can essentially train LLM with constant LR and just check the EMA weights to get its actual performance.

**Comment by tianyu-l** [comment]
We definitely should support EMA, just haven't got the bandwidth for it.

> In stable-LR phase, instead of checking eval on current weights, the averaged EMA weights can better measure the model's performance.

How does that help -- how is it making it closer to decay-LR regime?

**Comment by tianyu-l** [comment]
cc @fegin

**Comment by rakkit** [comment]
> How does that help -- how is it making it closer to decay-LR regime?

yep. exactly.  here is example from [nemotron-3-super](https://arxiv.org/abs/2604.12374) (page 49)

<img width="1660" height="1302" alt="Image" src="https://github.com/user-attachments/assets/61b8fe93-1584-4844-a8b8-b1b40b9760f9" />

and most recent works, for example [this one](https://arxiv.org/abs/2602.03702v1), are directing to just use EMA as model averaging, so the core code is like `EMA_weights = beta * EMA_weights + (1-beta) weights`,  at anytime we may only pay extra memory of `EMA_weights`. 

and i just realised we can build it on top of  `OptimizersContainer` and the rest seems just work out of box (state manage, parallelism,  sync/async DCP load/saving etc.. ).

**Comment by shuhuayu** [comment]
Thanks for the issue and pr. my understanding is that ema checkpoint is used for online checkpoint merging during wsd (warmup stable decay), any other use cases? for this use case, [nemotron super ](https://arxiv.org/abs/2604.12374) actually uses offline merging, and recent [kimi k3](https://arxiv.org/pdf/2607.24653) section 3.2 says that cosine decay is consistently better than wsd. but i think you mentioned that wsd is good for stop then pick up training, that can be a useful use case.

**Comment by rakkit** [comment]
yes, it is mainly for WSD or constant LR training (scheduler-free). and its more flexible to cut/extend training steps.

I think EMA version and offline version are more or less equivalent, we can go one of them.
